File: src/user_interface.cpp
/**
 * \file user_interface.cpp
 * \brief This node provides a user interface for controlling turtles in the turtlesim environment.
 * \author Rubin Khadka Chhetri
 * \version 1.0.0
 * \date 22/02/2025
 *
 * Publishes to:
 * - /turtle1/cmd_vel (geometry_msgs/Twist): Velocity commands for turtle1.
 * - /turtle2/cmd_vel (geometry_msgs/Twist): Velocity commands for turtle2.
 *
 * Services:
 * - /spawn (turtlesim/Spawn): Spawns a new turtle in the turtlesim environment.
 *
 * Description:
 * This node allows the user to control either turtle1 or turtle2 by entering linear and angular velocities.
 * The user can also spawn a new turtle (turtle2) at a predefined location.
 */

#include <ros/ros.h>
#include <std_msgs/String.h>
#include <turtlesim/Pose.h>
#include <turtlesim/Spawn.h>
#include <geometry_msgs/Twist.h>
#include <iostream>
#include <unistd.h>
#include <limits>

// Global variables to store linear and angular velocities
float linear_x, angular_z;

/**
 * \brief Main function for the turtle_user_interface node.
 * \param argc Number of command-line arguments.
 * \param argv Command-line arguments.
 * \return 0 on successful completion.
 *
 * This function initializes the ROS node, spawns a new turtle (turtle2), and provides a user interface
 * for controlling either turtle1 or turtle2 by publishing velocity commands.
 */

int main(int argc, char **argv) {
    // Initialise ROS
    ros::init(argc, argv, "turtle_user_interface");
    ros::NodeHandle nh;

    // Initialise service clients
    ros::ServiceClient client_spawn = nh.serviceClient<turtlesim::Spawn>("/spawn");
    turtlesim::Spawn spawn_srv;

    // Spawn Turtle
    spawn_srv.request.x = 5.0;
    spawn_srv.request.y = 2.0;
    spawn_srv.request.theta = 0.0;
    spawn_srv.request.name = "turtle2";
    client_spawn.call(spawn_srv);

    // Initialize publishers
    ros::Publisher pub_turtle1 = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
    ros::Publisher pub_turtle2 = nh.advertise<geometry_msgs::Twist>("/turtle2/cmd_vel", 10);

    std::string turtle_name;

    while (ros::ok()) {
        // User input
        std::cout << "Enter the turtle you want to control (turtle1 or turtle2): ";
        std::cin >> turtle_name;

        // Check if the turtle name is valid
        if (turtle_name != "turtle1" && turtle_name != "turtle2") {
            std::cout << "Invalid turtle name. Please enter 'turtle1' or 'turtle2'.\n";
            continue;
        }

        // Get the velocities
        std::cout << "Enter the linear velocity x (between -5 and 5): ";
        while (!(std::cin >> linear_x) || linear_x < -5 || linear_x > 5) { // Check range and input validity
            std::cout << "Invalid input. Please enter a linear velocity between -5 and 5: ";
            std::cin.clear(); // Clear error
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        }

        std::cout << "Enter the angular velocity z (between -5 and 5): ";
        while (!(std::cin >> angular_z) || angular_z < -5 || angular_z > 5) { // Check range and input validity
            std::cout << "Invalid input. Please enter an angular velocity between -5 and 5: ";
            std::cin.clear(); // Clear error
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        }

        // Create geometry msgs to move the turtle
        geometry_msgs::Twist turtle_vel;
        turtle_vel.linear.x = linear_x;
        turtle_vel.angular.z = angular_z;

        // Turtle1
        if (turtle_name == "turtle1") {
            pub_turtle1.publish(turtle_vel);
            ros::Duration(1.0).sleep();
            turtle_vel.linear.x = 0.0;
            turtle_vel.angular.z = 0.0;
            pub_turtle1.publish(turtle_vel);
        }
        // Turtle2
        else if (turtle_name == "turtle2") {
            pub_turtle2.publish(turtle_vel);
            ros::Duration(1.0).sleep();
            turtle_vel.linear.x = 0;
            turtle_vel.angular.z = 0;
            pub_turtle2.publish(turtle_vel);
        }

        ros::spinOnce();
    }

    ros::shutdown();
    
    return 0;
}

File: scripts/user_interface.py
#!/usr/bin/env python3

import rospy
from turtlesim.srv import Spawn
from turtlesim.msg import Pose
from geometry_msgs.msg import Twist


def check_if_turtle2_exists():
    """Checks if turtle2 exists by subscribing to /turtle2/pose."""
    turtle2_exists = False

    def pose_callback(msg):
        nonlocal turtle2_exists
        turtle2_exists = True  # Message received; turtle2 exists

    # Subscribe to /turtle2/pose topic
    rospy.Subscriber('/turtle2/pose', Pose, pose_callback)

    # Short delay, wait for message
    timeout_time = rospy.Time.now() + rospy.Duration(1.0)  # 1-second timeout
    while rospy.Time.now() < timeout_time and not rospy.is_shutdown():
        rospy.sleep(0.1)  # Sleep in small increments

    return turtle2_exists


def spawn_turtle():
    """Spawns turtle2 if it doesn't exist."""
    rospy.loginfo("Spawning Turtle2.")
    try:
        rospy.wait_for_service('/spawn', timeout=5)
        spawn_srv = rospy.ServiceProxy('/spawn', Spawn)
        spawn_srv(5.0, 2.0, 0.0, "turtle2")
    except rospy.ROSException as e:
        rospy.logerr(f"Spawn service not available: {e}")
    except rospy.ServiceException as e:
        rospy.logerr(f"Failed to spawn turtle2: {e}")


def main():
    # Initialize ROS node
    rospy.init_node('turtle_user_interface', anonymous=True)

    # Publishers for turtle1 and turtle2
    pub_turtle1 = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)
    pub_turtle2 = rospy.Publisher('/turtle2/cmd_vel', Twist, queue_size=10)

    # Check if turtle2 exists and spawn if necessary
    if check_if_turtle2_exists():
        rospy.loginfo("Turtle2 already exists.")
    else:
        spawn_turtle()

    rate = rospy.Rate(10)  # Loop rate

    while not rospy.is_shutdown():
        # Prompt user for turtle selection
        turtle_name = input("Enter the turtle you want to control (turtle1 or turtle2): ")

        # Validate turtle name
        if turtle_name not in ["turtle1", "turtle2"]:
            print("Invalid turtle name. Please enter 'turtle1' or 'turtle2'.")
            continue

        # Get linear velocity with range validation
        while True:
            try:
                linear_x = float(input("Enter the linear velocity (between -5 and 5): "))
                if -5 <= linear_x <= 5:
                    break
                else:
                    print("Invalid input. Linear velocity must be between -5 and 5.")
            except ValueError:
                print("Invalid input. Linear velocity must be a number.")

        # Get angular velocity with range validation
        while True:
            try:
                angular_z = float(input("Enter the angular velocity (between -5 and 5): "))
                if -5 <= angular_z <= 5:
                    break
                else:
                    print("Invalid input. Angular velocity must be between -5 and 5.")
            except ValueError:
                print("Invalid input. Angular velocity must be a number.")

        # Create and populate Twist message
        turtle_vel = Twist()
        turtle_vel.linear.x = linear_x
        turtle_vel.angular.z = angular_z

        # Publish to the selected turtle
        if turtle_name == "turtle1":
            pub_turtle1.publish(turtle_vel)
        elif turtle_name == "turtle2":
            pub_turtle2.publish(turtle_vel)

        # Move the turtle for 1 second
        rospy.sleep(1.0)

        # Stop the turtle
        turtle_vel.linear.x = 0.0
        turtle_vel.angular.z = 0.0
        if turtle_name == "turtle1":
            pub_turtle1.publish(turtle_vel)
        elif turtle_name == "turtle2":
            pub_turtle2.publish(turtle_vel)

        rate.sleep()


if __name__ == "__main__":
    main()
